### PR TITLE
Introduce `useEvent` hook

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

Introducing a hook that fires on event emitter's (e.g. native modules and shared objects) events and returns the event object of the most recently emitted event (or the initial value).

# How

Added `useEvent` hook that uses `addListener` under the hood and re-renders whenever a new event comes in.

# Test Plan

I tried to use the hook and the video player object and the event payload is property typed 👇 and updates whenever the source changes.

Screenshot 2024-06-20 at 11 05 54

### CODE REVIEW

**Comment by reichhartd** [comment]
Is the hook intended to standardise existing event listener hooks, such as [`useLowPowerMode`](https://github.com/expo/expo/blob/28aea768db84be0b81ef9675c2e91d3a726c6372/packages/expo-battery/src/Battery.ts#L243-L253) in Expo Battery, and make it easier to enable new ones in packages (for example [`useClipboard`](https://github.com/expo/expo/blob/28aea768db84be0b81ef9675c2e91d3a726c6372/packages/expo-clipboard/src/Clipboard.ts#L212-L227))?

**Comment by tsapeta** [comment]
@reichhartd In the long-term, yes. For now the main purpose is to provide an API for hooks that will be used in shared objects, e.g. the video player from `expo-video` and the database in `expo-sqlite`.
Eventually, custom listener wrappers like `addClipboardListener` will also be replaced by `NativeModule.addListener(...)`. What do you think about that?

**Comment by reichhartd** [comment]
@tsapeta Yep, sounds good and makes sense to replace the custom listener wrappers with a standardised method.
I asked because I was working on a draft for a function [`createEventEmitterHook`](https://github.com/reichhartd/expo/pull/2/files) a few weeks ago, and whether a similar function is also scope of this pull request.

File: packages/expo/src/hooks/useEvent.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  ? Parameters<TEventListener>[0]
+  : Parameters<TEventListener>[0] | TInitialValue | null;
+
+/**
+ * React hook that listens to events emitted by the given object. The returned value is an array of event parameters
+ * that get updated whenever a new event is dispatched.
+ * @param emitter An object that emits events, e.g. a native module or shared object or an instance of [`EventEmitter`](#eventemitter).
+ * @param event Name of the event to listen to.
+ * @param initialValue An array of event parameters to use until the event is called for the first time.
+ * @returns An array of arguments passed to the event listener.
```
Comment (Kudo) [inline packages/expo/src/hooks/useEvent.ts:38]: nit: maybe add `@example`

**Review by Kudo — APPROVED** [review]

File: packages/expo-modules-core/src/ts-declarations/EventEmitter.ts [hunk 2]
```diff
@@ -23,6 +23,15 @@ export type EventSubscription = {
  * Its implementation is written in C++ and common for all the platforms.
  */
 export declare class EventEmitter<TEventsMap extends EventsMap = Record<never, never>> {
+  /**
+   * A dummy private property with the given generic type. It is required for TypeScript to correctly infer this subtype.
+   * E.g. `useEvent` would not be able to infer the events map which results in accepting any string as the event name.
+   * @hidden
+   * @private
```
Comment (Simek) [inline packages/expo-modules-core/src/ts-declarations/EventEmitter.ts:30]: We can use only one of the params, we treat them as an aliases for docs rendering:
* https://github.com/expo/expo/blob/cae3c598a6bbf444f6b0eaad2acf152d0b9f4369/guides/Expo%20Documentation%20Writing%20Style%20Guide.md?plain=1#L330-L331

File: packages/expo/src/hooks/useEvent.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  TInitialValue,
+> = TInitialValue extends Parameters<TEventListener>[0]
+  ? Parameters<TEventListener>[0]
+  : Parameters<TEventListener>[0] | TInitialValue | null;
+
+/**
+ * React hook that listens to events emitted by the given object. The returned value is an array of event parameters
+ * that get updated whenever a new event is dispatched.
+ * @param emitter An object that emits events, e.g. a native module or shared object or an instance of [`EventEmitter`](#eventemitter).
+ * @param event Name of the event to listen to.
```
Comment (Simek) [inline packages/expo/src/hooks/useEvent.ts:36]:
```suggestion
 * @param eventName Name of the event to listen to.
```

Name match.

**Review by lukmccall — APPROVED** [review]

File: packages/expo/src/hooks/useEvent.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+ * @param initialValue An array of event parameters to use until the event is called for the first time.
+ * @returns An array of arguments passed to the event listener.
+ * @example
+ * ```tsx
+ * import { useEvent } from 'expo';
+ * import { VideoPlayer } from 'expo-video';
+ *
+ * export function PlayerStatus({ videoPlayer }: { videoPlayer: VideoPlayer }) {
+ *   const playerStatus = useEvent(videoPlayer, 'statusChange', videoPlayer.status);
+ *   return <Text>{`Player status: ${playerStatus}`}</Text>;
```
Comment (amandeepmittal) [inline packages/expo/src/hooks/useEvent.ts:46]:
```suggestion
 *
 *   return <Text>{`Player status: ${playerStatus}`}</Text>;
```

One tiny suggestion to add a newline before return statements for easier readability (as this is the common pattern when writing React code).

File: packages/expo/src/hooks/useEvent.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  TEventListener extends AnyEventListener,
+  TInitialValue,
+> = TInitialValue extends Parameters<TEventListener>[0]
+  ? Parameters<TEventListener>[0]
+  : Parameters<TEventListener>[0] | TInitialValue | null;
+
+/**
+ * React hook that listens to events emitted by the given object. The returned value is an array of event parameters
+ * that get updated whenever a new event is dispatched.
+ * @param emitter An object that emits events, e.g. a native module or shared object or an instance of [`EventEmitter`](#eventemitter).
```
Comment (amandeepmittal) [inline packages/expo/src/hooks/useEvent.ts:35]:
```suggestion
 * @param emitter An object that emits events. For example, a native module or shared object or an instance of [`EventEmitter`](#eventemitter).
```

Let's avoid using [Latin abbreviations as per our writing style guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#abbreviations)

**Review by Simek — APPROVED** [review]

**Review by amandeepmittal — APPROVED** [review]

**Pull request merged**